Improve performance of opam update/init by changing the structure of the internal http opam repositories (use the tar.gz as-is)

kit-ty-kate opened the pull request:
Fixes #5741
Fixes https://github.com/ocaml/opam/issues/5648
Fixes https://github.com/ocaml/opam/issues/5484
~~Fixes https://github.com/ocaml/opam/issues/5346~~
Fixes https://github.com/ocaml/opam/issues/5559
Fixes https://github.com/ocaml/opam/issues/3050
Fixes #6974 
cc @hannesm to check if it works for conex (2.1 worked with tar.gz files already so i'm not too scared about breakages)

### Reasoning

#5741 shows that assumptions that hold true "most of the time" on some unix platforms such as "it is ok to scan a large tree structure of files and directories", don't hold true on other platforms. Systems such as Windows, network filesystems, busy shared servers whose disk is being constantly used, harddrives, … suffer from this.

In opam we can have 3 types of repositories:
1. HTTP (the default): where we download a tar.gz
2. VCS (aka. mostly git these days): where we use the vcs command line to get the files
3. local/ssh: where we either use our own copy primitives or use rsync

Out of the three, the most critical for first time users is the first one. It is also the one that suffer the most from these issues as currently we:
1. untar it
2. diff with the current repository (used for conex and for https://github.com/ocaml/opam/pull/6614)
3. patch the changed files
4. remove the directory
5. rescan the whole repository

VCS do not have step 4. Steps 1, 2 and 3 are builtin and heavily optimized. Is left only step 5 which should be improved by https://github.com/ocaml/opam/pull/6614 and for which we can improve further later by using `git cat-file` or even parse PACK files using `ocaml-git`.

Local/ssh repositories are the ones left a bit with very few things we can do about them. https://github.com/ocaml/opam/pull/5966 should help, but beyond, maybe we might want to require that people use git even for local repositories.

For HTTP though, the untarring (which takes 1+ minute) is the main issue. Thus this here PR.

### Design decisions

Instead of untarring we simply use the tar.gz as-is and use `ocaml-tar` to read it on the fly.
The new update steps are:
1. diff the two tar.gz
2. remove the old tar.gz
3. move the new tar.gz in its place
4. scan what has changed (required https://github.com/ocaml/opam/pull/6614)

Given the ubiquity of the use of `OpamFilename.Dir.t` to mean both any random directory and a repository directory, i chose to first abstract over it in a new `OpamRepositoryRoot` module, and work with the help of the type checker from there. Its interface help see what are the actions that opam does on repositories. While i'd rather keep them, the `Tar` and `Dir` submodules can be removed when everything is done.

The `REPOSITORYTARRING` environment variable is removed by this work, given the repositories are tarred already.

I had simplify `opam var pkg:opamfile` for this work. Previously it would point to the file in the repository. However this isn't what it's supposed to be doing. Instead it should point to the `<switch>/.opam-switch/packages/` directory which actually reflects the opam file that was used to installed. Otherwise the opam file can change between before and after the user has called `opam update` etc.

### TODO

There are a number of `assert false (* TODO *)` in this draft PR. Those are to be fixed before undrafting but i felt reasonably confident with the rest of them to open this draft PR in this state to put more eyes on this work and to increase my self-motivation.

- The main thing to do is to do the diff function between two tar files and between a directory and a tar.
- The other thing is to fill `OpamRepositoryState.get_repo_files`: a function which extracts a limited number of files from the tar.gz to a new cache directory.

Some of these changes should probably be extracted to separate PRs but let's do that at the end when we have something that actually works.

While early form of this work started a year and a half ago, i believe the crust left over from that time should be minimal, after 6 different branches. The final rebase and split into smaller PRs shouldn't be too painful.

### Future work

In the future we can use `ocaml-tar` that we now depend on to replace some of the uses of the `tar` command. This should allow us to have better behaviours with things like symlinks on windows or even add new features such as excluding some directories (see https://github.com/ocaml/ocaml/pull/14152).

As mentioned above we can also improve local and git/vcs repositories with or without `ocaml-git`.


---
Queued PRs
* [x] ~~#6871~~
* [x] #6913
* [x] #6914 
* [x] #6915 
* [x] #6917 
* [x] #6680
* [x] #6934
* [x] #6945
* [x] #6949
* [x] #6941
* [x] #6939
* [x] #6970
* [x] #6971
* [x] #6998 
* [x] #7002


A run of reftests with `OPAMREPOSITORYTARRING` enabled for all tests can be found on top of [rjbou/repotar-reftest-repotarring-enabled](https://github.com/rjbou/opam/tree/repotar-reftest-repotarring-enabled)

hannesm commented:
Hey, thanks for your work on that. Since you asked me directly

> cc @hannesm to check if it works for conex

> Instead of untarring we simply use the tar.gz as-is and use ocaml-tar to read it on the fly.
> The new update steps are:
> 
>    diff the two tar.gz
>    remove the old tar.gz
>    move the new tar.gz in its place

This should be fine from the design point of view for conex.

Conex will need to interject between step 0 (you downloaded the tarball) and 2 (remove the old tar.gz). Currently, conex requires a diff file on disk, and the old repository as directory. But we can revise that interface, and conex could as well work on two tar.gz (and/or on two directories).

I guess you have a clear understanding of the update process currently, and since you mention the different kinds (http, git, local) -- maybe we should re-think how opam and conex should interact to avoid burden paid by people not using conex, and avoid the burden of duplicating computations in both opam and conex. The latter may need to include conex as a library into opam.

I'm away for the next 10 days (back on August 10th), but am happy to discuss this afterwards - esp.since I plan to revive my work on conex thereafter.

hannesm commented:
To be more precise, given that
- (a) local repositories (rsync) aren't really worth to verify (under the assumption that whoever has access to the repository can as well install arbitrary packages) [which may be revisited if there's the NFS use case or a shared server]
- (b) VCS (git): we could do the `git fetch` and provide conex with the local repository and the commit that the update should be to
- (c) http: provide old and new tarball

Then conex could do what is needed (compute the set of changed opam files, verify signatures; exit 0 on success); and could even report back the set of changed files to opam (I suspect this is what #6614 depends on) - using a file, or a socket, or if integrated with opam, this will be much simpler (using shared memory).

For opam itself, I guess that #6349 and #5553 will improve a lot of updates already.

kit-ty-kate commented:
Some update: I've been splitting this PR into several smaller ones to make working on it collaboratively, rebasing it and cleaning it up, easier.

So far:
* #6658, https://github.com/ocaml/opam/pull/6660, https://github.com/ocaml/opam/pull/6661, https://github.com/ocaml/opam/pull/6662 were merged
* https://github.com/ocaml/opam/pull/6659, https://github.com/ocaml/opam/pull/6679, https://github.com/ocaml/opam/pull/6680 are being worked on and will be merged in order

Once these done, i expect this PR to be fairly small, self-contained and easier to review.

rjbou commented on src/repository/opamRepositoryBackend.ml:
```diff
@@ -126,7 +126,20 @@ let strip_repo_suffix patch =
   in
   {patch with operation}
 
-let get_diff parent_dir dir1 dir2 =
```
`get_diff`is historically in `OpamRepositoryBackend`, but given the functions added, its use in another PR for pinning stuff, i think it should be moved. Maybe a `core/opamDiff.ml`?

rjbou replied:
In end up a single function in last versions of the code.

dinosaure commented on src/repository/opamTar.ml:
```diff
@@ -0,0 +1,121 @@
+              Tar.Header.make ~file_mode:0 ~mod_time:0L ~user_id:0 ~group_id:0
+                path (Int64.of_int (String.length content))
+            in
+            let data = fun () -> Tar.return (Ok (Some content)) in
```
Here, the given stream must emit `None` at some points. Something like:
```ocaml
let data =
  let closed = ref false in
  fun () -> match !closed with
    | false -> closed := true; Tar.return (Ok (Some content))
    | true -> Tar.return (Ok None) in
...
```

Should fix you infinite loop.

dinosaure commented on src/repository/opamTar.ml:
```diff
@@ -0,0 +1,119 @@
+
+  let add ~fname ~content (fd, t) = (fd, Map.add fname content t)
+
+  let remove fname (fd, t) = (fd, Map.remove fname t)
+
+  let write (fd, t) =
+    let to_buffer buf t =
+      let rec run : type a. Buffer.t -> (a, 'err, _) Tar.t -> a = fun buf -> function
+        | Tar.Write str -> Buffer.add_string buf str
+        | Tar.Read _ | Tar.Really_read _ | Tar.Seek _ | Tar.High _ -> assert false
+        | Tar.Return (Ok value) -> value
+        | Tar.Return (Error _) -> failwith "something went wrong"
+        | Tar.Bind (x, f) -> run buf (f (run buf x))
+      in
+      run buf t
+    in
+    let entries =
+      let x =
+        Map.fold (fun path content acc ->
+            let hdr =
+              Tar.Header.make ~file_mode:0 ~mod_time:0L ~user_id:0 ~group_id:0
+                path (Int64.of_int (String.length content))
+            in
+            (Some Tar.Header.Ustar, hdr, fun () -> Tar.return (Ok (Some content))) :: acc)
+          t []
+      in
+      let r = ref x in
+      fun () ->
+        match !r with
+        | [] -> Tar.return (Ok None)
+        | x::xs -> r := xs; Tar.return (Ok (Some x))
+    in
+    let t = Tar.out ~level:Ustar entries in
+    let t = Tar_gz.out_gzipped ~level:4 ~mtime:0l Gz.Unix t in
+    let buf = Buffer.create 10_485_760 in
+    to_buffer buf t;
+    let str = Buffer.contents buf in
+    let _ : int = Unix.write_substring fd str 0 (String.length str) in
+    ()
```
To be precise (and this is exactly as you requested), here we could use `Unix.write_string` directly rather than buffering and writing the archive in one go (which would remove the atomicity but reduce memory usage; another approach is to write to `/tmp` and use `Unix.rename`).

The basic idea is to decouple the I/O from the generation of a `tar.gz` file so that the user can decide how to actually write it.

A final suggestion would also be to generate a `string Seq.t` at the end if you prefer.

rjbou committed:
OpamFile: add safe_read_from_string

rjbou committed:
OpamTar: update create to permit to create flat directory archives

kit-ty-kate committed:
OpamSystem.directories_with_links: add argument to remove VCS directories

Co-authored-by: Raja Boujbel <raja.boujbel@ocamlpro.com>

rjbou committed:
OpamPatch: Add a configurator to be able to patch either on system files or in tar files

rjbou committed:
OpamFilename.Unix: expose basename & dirname

rjbou committed:
OpamFilename.Unix: add remove_prefix

rjbou committed:
OpamRepositoryRoot: add read_file

rjbou committed:
reftest: add repo set-url test

rjbou committed:
OpamFilename.Unix: add root_dir

rjbou committed:
OpamFilename.Unix.Dir: add basename & dirname

rjbou committed:
OpamRepositoryRoot: add remove_prefix & remove_prefix_dir

rjbou committed:
reftest: add a test in repository-http for switching from directory to archive representation

rjbou committed:
OpamFile: add location information for 'read_from_string' functions

Add also debug logging with level 3, that uses that new location
information

rjbou committed:
Add conditional upgrade opam root for 2.6

This opam root upgrade is triggered only if there is an already present
archive repository, generated from old OPAMREPOSITORYTARRING usage, or
2.1 (that does it automatically). It reconstruct that archive in the new
version of repository storage archive (no root directory).

rjbou commented on src/repository/opamTar.ml:
```diff
@@ -0,0 +1,235 @@
+      [Unix.O_CREAT; Unix.O_TRUNC; Unix.O_WRONLY] 0o640
+  in
+  Fun.protect ~finally:(fun () -> Unix.close fd) @@ fun () ->
+  let files = OpamFilename.rec_files dir in
```
```suggestion
  let files = OpamFilename.rec_files ~except_vcs:true dir in
```

rjbou commented on src/core/opamPatch.ml:
```diff
@@ -322,87 +322,165 @@ let translate_patch ~dir orig corrected =
-        Option.iter (OpamSystem.write (file^".orig")) content;
-        OpamSystem.write (file^".rej") (Format.asprintf "%a" Patch.pp diff);
+        (* TAR TODO : write somewhere else ?
+           Option.iter (write (file^".orig")) content;
```
add an function in configurator fro writing rejected files : write on disk, no-op on archive

rjbou commented on src/core/opamPatch.ml:
```diff
@@ -322,87 +322,165 @@ let translate_patch ~dir orig corrected =
 
+(* Patch configurator *)
+
+module type PATCH_CONF = sig
```
rename on `FS_ABSTR`

rjbou commented on src/repository/opamRepositoryRoot.ml:
```diff
@@ -56,62 +56,182 @@ module Dir = struct
+
+  let archives : (OpamHash.t, string OpamFilename.Unix.Map.t) Hashtbl.t = Hashtbl.create 8
+  let unload_repo_tars () = Hashtbl.clear archives
+
```
extract this memoization to a draft PR, we'll see later if we need it or no

rjbou commented on src/client/opamRepositoryCommand.ml:
```diff
@@ -251,18 +251,15 @@ let update_with_auto_upgrade rt repo_names =
-               | OpamRepositoryRoot.Dir x -> x
+             let repo_root = OpamRepositoryState.get_repo_root rt r in
+             OpamRepositoryRoot.on_dir (fun dir ->
+                 OpamAdminRepoUpgrade.do_upgrade (OpamRepositoryRoot.Dir.of_dir dir))
```
Open an issue for updating do_upgrade to perform upgrade on tar directly

rjbou replied:
opened: #6994

rjbou commented on src/state/opamPinned.ml:
```diff
@@ -310,3 +310,32 @@ let orig_opam_file st name opam =
+      (* TAR TOQUESTION : some comments to check
+         it is better to fold over the full repo only once and the look again
+         on smaller lists instead of looking twice over all repo where there is
+         no locked file *)
```
remove comment, and update the function (maybe in the rewrite commit ?) to no longer check for `*.opam` & `*.opam.locked` when it is looking into a repository

rjbou commented on src/repository/opamRepositoryRoot.ml:
```diff
@@ -56,62 +56,182 @@ module Dir = struct
+    in
+    OpamTar.patch ~allow_unclean patch_source tar
+
+  let extract_files cond t =
```
s/extract_files/filter_files

rjbou commented on src/repository/opamRepositoryRoot.ml:
```diff
@@ -56,62 +56,182 @@ module Dir = struct
+        contents x
+    | None ->
+      let result, map =
+        OpamTar.fold_reg_files (fun (acc, map) file content ->
```
When we retrieve an archive from outside, we need to perform a pass to if some paths are canonical:
* `^./` : delete the ./
* `..` at any place : fail
* `/` : fail

Add the check in OpamTar.fold_reg_files
see #6924 canonical function

rjbou commented on src/state/opamFileTools.ml:
```diff
@@ -1612,6 +1612,13 @@ let read_repo_opam_t ~repo_name ~repo_root dir file content xfs =
+  let file = OpamFilename.Unix.to_filename file in
+  let dir = OpamFilename.Unix.Dir.to_dir dir in
+  let repo_root = OpamRepositoryRoot.Tar repo_root in
+  let xfs = OpamFilename.Unix.Map.map (fun c -> lazy c) xfs in
```
Use Lazy.from_val

rjbou commented on src/state/opamRepositoryState.ml:
```diff
@@ -102,6 +103,30 @@ let get_repo_files rt name dir =
+                    (OpamRepositoryName.to_string name)
+                  // OpamFilename.Unix.to_string filename))
+              (OpamRepositoryRoot.Tar.to_string tar);
+            content)
```
remove in final, kept ftm for debug

rjbou commented on src/state/opamRepositoryState.ml:
```diff
@@ -167,6 +167,56 @@ let read_package_opam_tar ~repo_name ~repo_root package_dir filename content ext
       (OpamFilename.Unix.to_string OpamFilename.Unix.Op.(package_dir // "opam"));
     None
 
+let load_raw_opams_and_aux_from_tar _repo_name tar =
```
add comment for the different steps

rjbou commented on src/state/opamRepositoryState.ml:
```diff
@@ -167,6 +167,56 @@ let read_package_opam_tar ~repo_name ~repo_root package_dir filename content ext
+              if OpamFilename.Unix.starts_with dir filename then
+                raise (Found (dir, value))
+            ) acc;
+          (* TAR TOQUESTION add a skipping message ? *)
```
remove comment & add newline at the end of the function

rjbou commented on src/client/opamListCommand.ml:
```diff
@@ -875,6 +875,16 @@ let info st ~fields ~raw ~where ?normalise ?(show_empty=false)
+               rdir
+           else
            OpamFilename.Dir.to_string
              OpamRepositoryRoot.Dir.Op.(repo_dir / rdir)
```
+ indent

rjbou commented on src/state/opamFormatUpgrade.ml:
```diff
@@ -1152,6 +1152,53 @@ let v2_2 = OpamVersion.of_string "2.2"
+              () file;
+            false
+          with Found is_repo_or_package_dir -> not is_repo_or_package_dir))
+    (OpamFilename.files (root / OpamRepositoryPathName.repo_d))
```
fold over repo list instead of files

rjbou commented on tests/reftests/opamroot-versions.test:
```diff
@@ -3567,3 +3567,305 @@ STATE                           Switch state loaded in 0.000s
 # Name            # Installed # Synopsis
 i-am-package      2           One-line description
 i-am-sys-compiler 2           One-line description
+### rm -rf _opam
```
add also a test with an archive present to see the hard upgrade

rjbou replied:
done in repository test. It too complicated (and duplicates a lot) to add it here

rjbou committed:
OpamTar: update create to permit not embedding VCS files

rjbou committed:
OpamFilename.Unix: add to_relative_canonical

rjbou committed:
OpamTar: when opening an archive, check paths internally

rjbou committed:
OpamFilename.Unix: when converting to filename or dir, put back windows dir sep

rjbou committed:
reftests: enable path substitution on 'packages/' partial paths & co

rjbou committed:
reftests: add repository test format change from dir to tar and vice versa

rjbou commented on tests/reftests/repository.test:
```diff
@@ -967,3 +966,577 @@ Now run 'opam upgrade' to apply any package updates.
+eval-variables: [
+  [
+    sys-ocaml-version
+    ["ocamlc" "-vnum"]
```
```suggestion
### head -2 OPAM/config
opam-version: "2.0"
opam-root-version: "2.5.0"
```

rjbou commented on tests/reftests/repository.test:
```diff
@@ -967,3 +966,577 @@ Now run 'opam upgrade' to apply any package updates.
+eval-variables: [
+  [
+    sys-ocaml-version
+    ["ocamlc" "-vnum"]
```
```suggestion
### head -2 OPAM/config
opam-version: "2.0"
opam-root-version: "2.5.0"
```

kit-ty-kate committed:
Change internal storage of repositories in opam root from only directories to directories and tar gz archives

By default, http repositories will be stored as archives. It is possible
to enable non-VCS repositories as archive by setting
`OPAMREPOSITORYTARRING`.

Co-authored-by: Raja Boujbel <raja.boujbel@ocamlpro.com>

rjbou committed:
OpamRepositoryRoot: add documentation for patch

rjbou committed:
reftests: add repository test case that triggers opam root upgrade to 2.6 with OPAMREPOSITORYTARRING

rjbou commented:
The main comment will need a rewrite, and maybe the main commit message

rjbou commented on src/core/opamFilename.mli:
```diff
@@ -442,4 +442,8 @@ module Unix : sig
 
+  (** Return [Error err] if path is absolute or contains '..'
+      Return [Ok filename] where trailing '/' are removed and './' at the
+      beginning *)
```
```suggestion
      Return [Ok filename] where trailing '/' and './' at the beginning
      are removed *)

kit-ty-kate commented on src/repository/opamTar.ml:
```diff
@@ -195,12 +205,50 @@ let create tar dir =
+  let on_rejection _ _ _ = ()
+  let ext file ext = OpamFilename.Unix.add_extension file ext
+  let write file content target = Tar.add file content target
+  let exists file = Tar.exists file
```
```suggestion
  let ext = OpamFilename.Unix.add_extension
  let write = Tar.add
  let exists = Tar.exists
```

kit-ty-kate commented on src/repository/opamTar.ml:
```diff
@@ -195,12 +205,50 @@ let create tar dir =
+  let exists file = Tar.exists file
+  let exists_dir _file _target = false
+  let read file = Tar.read file
+  let remove file = Tar.remove file
```
```suggestion
  let read = Tar.read
  let remove = Tar.remove
```

kit-ty-kate commented on src/core/opamPatch.mli:
```diff
@@ -19,18 +19,95 @@
+
+  (* For directories, we need to add an end slash *)
+  val end_slash : root -> root
+
```
This function only makes sense in get_path
```suggestion
```

kit-ty-kate commented on src/repository/opamTar.ml:
```diff
@@ -204,3 +204,36 @@ let create tar dir =
+  let file_to_string = OpamFilename.Unix.to_string
+  let end_slash = Fun.id
+  let get_path ~fail:_ _target =
+    (* check if escapable ? *)
```
yes i think check_canonical should be used here

rjbou commented on src/repository/opamTar.ml:
```diff
@@ -64,6 +64,16 @@ let run archive =
+                    (OpamFilename.Unix.to_string filename) err
+    in
+    f acc filename content
+  in
```
An update in the documentation of the module is needed to highlight that check & "rewrite"

kit-ty-kate commented on src/core/opamFilename.mli:
```diff
@@ -420,9 +427,31 @@ module Unix : sig
+
+  (** Adds a dot and the given file extension *)
+  val add_extension: t -> string -> t
+
```
This function isn't used anywhere (it's only used was in PatchFS but its result is ignored)
```suggestion
```

rjbou replied:
The result is stored in the archive, which is not the best. Should we drop it, or as rejection, have a `FS.on_accept` that writes only on disk ?

kit-ty-kate replied:
Technically it could be used but in practice `OpamPatch.patch` is never called with `~allow_unclean:true` on a repository so it will never get used

kit-ty-kate replied:
Then i guess adding an `FS.on_unclean_accept` should be fine

kit-ty-kate commented on src/core/opamPatch.mli:
```diff
@@ -19,18 +19,95 @@
+  val ext : file -> string -> file
+
+  (* Write a [file] *)
+  val write : file -> string -> target -> target
```
```suggestion
  (* Write a [file] *)
  val write : file -> ?ext:string -> string -> target -> target
```

kit-ty-kate commented on src/core/opamPatch.ml:
```diff
@@ -322,87 +322,159 @@ let translate_patch ~dir orig corrected =
+      let patching, content = patch file (Some content) diff patching in
+      let patching = FS.write file content patching in
+      let patching =
+        if file1_exists && file1 <> (file2 : FS.file) then
```
`FS.file` is abstract. There is no guarantee on the structure of it so we need to use a proper equality function
```suggestion
        if file1_exists && not (FS.eq_file file1 file2) then
```

kit-ty-kate commented on src/core/opamPatch.ml:
```diff
@@ -322,87 +322,159 @@ let translate_patch ~dir orig corrected =
+    FS.get_path ~fail to_patch file
   in
-  let patch ~file content diff =
+ let patch file content diff patching =
```
```suggestion
  let patch file content diff patching =
```

kit-ty-kate reviewed (commented):
partial review up to c1094b255436285a65b8a64d74becdf80aec4001 (also partially)

rjbou commented on src/core/opamFilename.ml:
```diff
@@ -450,25 +450,48 @@ let link ?(relative=false) ~target ~link =
+    if not (OpamStd.String.is_prefix_of ~from:0 ~full:file dir) then
+      fail ();
+    file
+  let ext file ext = file ^ ext
```
```suggestion
  let ext file ext = file ^ "." ^ ext

kit-ty-kate replied:
This doesn't match the behaviour of its usage in `OpamPatch.patch_t`

kit-ty-kate commented on src/core/opamFilename.ml:
```diff
@@ -800,18 +816,83 @@ module Unix = struct
+    | ""::_ -> Error "begins with '/'"
+    (* ^./ : delete the ./ *)
+    | "."::elements -> check_parent_dir_and_remove_trailing_slashes elements
+    | [_] -> Ok filename
```
this ignores the `".."` case as well as `"./"`

kit-ty-kate commented on src/core/opamFilename.ml:
```diff
@@ -800,18 +816,83 @@ module Unix = struct
+    (* ^./ : delete the ./ *)
+    | "."::elements -> check_parent_dir_and_remove_trailing_slashes elements
+    | [_] -> Ok filename
+    | elements -> check_parent_dir_and_remove_trailing_slashes elements
```
i'd rather replace this with the version from https://github.com/ocaml/opam/pull/6924/changes/3f057a83fa29a3172cab54e1ad991a1412b170b2
```suggestion
  let to_relative_canonical_list x =
    if Filename.is_relative x then
      match String.split_on_char '/' x with
      | [] | ""::_::_ -> assert false
      | [""] -> [current_dir_name]
      | x ->
        List.filter (fun seg ->
            if seg = ".." then
              invalid_arg "'..' is forbidden"
            else
              seg <> "" && not (String.equal current_dir_name seg))
          x
    else
      invalid_arg "not relative"

  let check_relative_canonical filename =
    try Ok (String.concat dir_sep (to_relative_canonical_list filename))
    with Invalid_argument msg -> Error msg
```

kit-ty-kate commented on src/core/opamPatch.mli:
```diff
@@ -19,18 +19,95 @@
+     @param content is the file content, if it exists
+     @param diff is the patch that failed to apply
+  *)
+  val on_rejection : file -> string option -> Patch.t -> unit
```
```suggestion
  val on_unclean_rejection : file -> string option -> Patch.t -> unit
```

kit-ty-kate commented on src/repository/opamTar.ml:
```diff
@@ -195,12 +205,50 @@ let create tar dir =
+    Tar.remove_dir (OpamFilename.Unix.dirname file) target
+  let same_dirname ~src ~dst =
+    OpamFilename.Unix.dirname src
+    <> (OpamFilename.Unix.dirname dst : OpamFilename.Unix.Dir.t)
```
```suggestion
    OpamFilename.Unix.Dir.equal
      (OpamFilename.Unix.dirname src)
      (OpamFilename.Unix.dirname dst)
```

kit-ty-kate commented on src/core/opamFilename.ml:
```diff
@@ -450,25 +450,48 @@ let link ?(relative=false) ~target ~link =
+  let remove file _target = OpamSystem.remove_file file
+  let remove_dir file _target = OpamSystem.rmdir_cleanup (Filename.dirname file)
+  let same_dirname ~src ~dst =
+    Filename.dirname src <> (Filename.dirname dst : string)
```
```suggestion
    Filename.dirname src = (Filename.dirname dst : string)
```

kit-ty-kate commented on src/core/opamPatch.ml:
```diff
@@ -322,87 +322,159 @@ let translate_patch ~dir orig corrected =
+        if FS.same_dirname ~src ~dst then
+          FS.remove_dir src patching
+        else
+          patching
```
```suggestion
        if FS.same_dirname ~src ~dst then
          patching
        else
          FS.remove_dir src patching
```

kit-ty-kate commented on src/repository/opamRepositoryRoot.mli:
```diff
@@ -76,12 +115,32 @@ val dirname : t -> dirname
 val to_string : t -> string
 
+val remove_prefix: filename -> t -> filename
+val remove_prefix_dir: dirname -> t -> dirname
```
It would otherwise be inconsistent with `OpamFilename` etc
```suggestion
val remove_prefix: t -> filename -> filename
val remove_prefix_dir: t -> dirname -> dirname
```

kit-ty-kate commented on tests/reftests/run.ml:
```diff
@@ -664,6 +664,10 @@ let common_filters ?opam dir =
     Sed "${OPAMTMP}";
+    (* We need this sed to ensure that the line containing 'packages' is
+       rewritten with the good dir sep replacements *)
+    seq [ str "packages"; set "/\\" ],
```
why would the rest of the path only contains `/` but only the packages part would contain `\`?

rjbou replied:
The function `str_replace_path` handles the reste of the path

kit-ty-kate commented on src/client/opamArg.ml:
```diff
@@ -224,10 +224,10 @@ let environment_variables =
-       but have good caching in /tmp. However this is slower in the \
-       general case.";
+      "internally store all non-vcs repositories as tar.gz files (by default, \
+       http repositories are stored as archive).This can be much faster on \
```
```suggestion
       http repositories are stored as archive). This can be much faster on \
```

kit-ty-kate commented on src/client/opamCommands.ml:
```diff
@@ -2764,7 +2762,7 @@ let switch cli =
       ~extra_defaults:[
         cli_from cli2_2, "-",
-        "Switches back to the previous non-local switch (similar to $(b,git switch -)). \
+         "Switches back to the previous non-local switch (similar to $(b,git switch -)). \
```
```suggestion
        "Switches back to the previous non-local switch (similar to $(b,git switch -)). \
```

kit-ty-kate commented on src/format/opamRepositoryPath.mli:
```diff
@@ -104,7 +104,8 @@ module Make (I : PATH_REPR) : PATH
 
-val tar: dirname -> repository_name -> filename
+(** obsolete, no longer used *)
+val repo_tarring: dirname -> repository_name -> filename
```
If it's not used, why not simply remove it?
```suggestion
```

kit-ty-kate commented on src/repository/opamRepositoryRoot.mli:
```diff
@@ -76,12 +115,32 @@ val dirname : t -> dirname
+val remove_prefix_dir: dirname -> t -> dirname
+
+val is_tgz: t -> bool
+val is_dir: t -> bool
```
is_dir isn't used and is_tgz is only used to get the name so it would make more sense to simply get the name directly from OpamRepositoryRoot
```suggestion
val get_backend_name: t -> string
```

kit-ty-kate commented on src/repository/opamRepositoryRoot.mli:
```diff
@@ -90,3 +149,13 @@ val patch :
     [repo] file exists in the repository and [f] reads it *)
 val delayed_read_repo : t -> bool * (unit -> OpamFile.Repo.t)
+
+(** TAR TODOC remove directory and archive, if present *)
```
```suggestion
(** remove directory and archive, if present *)
```

kit-ty-kate commented on src/state/opamFileTools.ml:
```diff
@@ -1321,17 +1365,40 @@ let add_aux_files ?dir ?(files_subdir_hashes=false) opam =
     in
-    let files_dir =
-      OpamFilename.Op.(dir / OpamRepositoryPathName.files_d)
+    let files_dir = OpamFilename.Op.(dir / "files") in
```
```suggestion
    let files_dir = OpamFilename.Op.(dir / OpamRepositoryPathName.files_d) in
```

kit-ty-kate commented on tests/reftests/action-disk.test:
```diff
@@ -110,8 +110,12 @@ SYSTEM                          mkdir ${BASEDIR}/OPAM/repo/default/packages/main
-FILE(opam)                      Read ${BASEDIR}/OPAM/repo/default/packages/main-repo/main-repo.1/opam in 0.000s
+SYSTEM                          read ${BASEDIR}/OPAM/repo/default/packages/main-repo/main-repo.1/opam
+FILE(opam)                      Read packages/main-repo/main-repo.1/opam (out of ${BASEDIR}/OPAM/repo/default) in 0.000s
+SYSTEM                          read ${BASEDIR}/OPAM/repo/default/packages/main-repo/main-repo.1/files/x-file.main-repo.1
```
why are we suddenly reading the extra-files?

rjbou replied:
On loading, there is a checksum checks, extra files were directly read by `Digest.file`, while here, we read them and use  `Digest.string`

kit-ty-kate commented on tests/reftests/repository-layout.test:
```diff
@@ -147,15 +147,21 @@ RSTATE                          Cache found
+UPDATE                          Repository did not change: nothing to do.
+FILE(config)                    Read ${BASEDIR}/OPAM/config in 0.000s
+RSTATE                          LOAD-REPOSITORY-STATE @ ${BASEDIR}/OPAM
+RSTATE                          Cache found
```
I believe this is an artefact of the rebase over https://github.com/ocaml/opam/pull/6941
```suggestion
```

kit-ty-kate commented on tests/reftests/repository.test:
```diff
@@ -62,7 +62,6 @@ plain
 repos-config
 tarred.tar.gz
 ### opam install foo.3 -vv | grep '^\+' | sed-cmd test tar
```
```suggestion
### opam install foo.3 -vv | grep '^\+' | sed-cmd test
```

kit-ty-kate commented on master_changes.md:
```diff
@@ -71,6 +71,7 @@ users)
 ## Repository
   * No longer call tar tool to create archives, use tar library instead [#6945 @kit-ty-kate]
   * [BUG] Do not fail on directories named `opam` when scanning the `packages` directory of a repository during `opam repo add` or `opam init` (worked on subsequent `opam update`) [#6941 @kit-ty-kate @rjbou]
+  * Speedup repository operations on Windows & BSD by changing its storage in the opam root from plain directory to archive, for HTTP repositories, or non-VCS one if `OPAMREPOSITORYTARRING` is enabled [#6625 @rjbou @kit-ty-kate @arozovyk - fix #5741 #5648 #5484 #5559 #3050]
```
```suggestion
  * Speedup repository operations on certain file-systems (e.g. NTFS on Windows or IO constrained machines) by changing its storage in the opam root from plain directory to archive, for HTTP repositories, or non-VCS one if `OPAMREPOSITORYTARRING` is enabled [#6625 @rjbou @kit-ty-kate @arozovyk - fix #5741 #5648 #5484 #5346 #5559 #3050 #6974]
```

kit-ty-kate reviewed (commented):
lgtm modulo green CI otherwise

rjbou committed:
OpamRepositoryRoot: add string_of_backend

rjbou commented:
Should have resolved all suggestions. I've pushed twice (without and with rebase, to highlight the diff). Test branch is updated too.

kit-ty-kate commented on src/core/opamFilename.ml:
```diff
@@ -462,25 +462,51 @@ let link ?(relative=false) ~target ~link =
+    Option.iter (fun c -> write (file^".orig") c ()) content
+  let on_unclean_reject file content diff =
+    on_unclean_accept file content;
+    Option.iter (fun c -> write (file^".orig") c ()) content;
```
This line seems duplicated with the new call just above
```suggestion
```

rjbou commented:
Updated

kit-ty-kate commented:
Ignoring cygwin CI failure (fixed by #7007) and random mingw failure.

:tada:
